Sources don't compile with gcc10 (on Arch)

This seems to be the same kind of snag that we've recently encountered with purr-data, too, but which Jonathan has since fixed. With gcc10, while trying to link the pd-l2ork executable, I'm getting a lot of linkage errors of this kind:

~~~
[  381s] cd ../obj;  cc -Wl,-export-dynamic -lasound -lrt -ljack -lfftw3  -o ../bin/pd-l2ork d_ctl.o d_array.o d_delay.o d_filter.o d_math.o d_osc.o d_soundfile.o g_canvas.o g_graph.o g_text.o g_rtext.o g_array.o g_template.o g_io.o g_scalar.o g_traversal.o g_guiconnect.o g_readwrite.o g_editor.o g_clone.o g_undo.o g_all_guis.o g_bang.o g_mycanvas.o g_numbox.o g_toggle.o g_vumeter.o g_magicglass.o g_radio.o g_slider.o m_pd.o m_class.o m_obj.o m_atom.o m_memory.o m_binbuf.o m_conf.o m_glob.o m_sched.o s_main.o s_inter.o s_file.o s_print.o s_loader.o s_path.o s_entry.o s_audio.o s_midi.o s_utf8.o d_ugen.o d_arithmetic.o d_dac.o d_misc.o d_fft.o d_global.o d_resample.o x_arithmetic.o x_connective.o x_interface.o x_midi.o x_misc.o x_time.o x_acoustics.o x_net.o x_text.o x_gui.o x_list.o x_array.o x_preset.o x_scalar.o x_vexp.o x_vexp_if.o x_vexp_fun.o import.o s_midi_oss.o s_audio_oss.o s_audio_alsa.o s_audio_alsamm.o s_midi_alsa.o s_audio_jack.o d_fft_fftw.o d_fftroutine.o  -ldl -lm -lpthread -lfftw3f -lasound -ljack -ljack
[  382s] /usr/bin/ld: g_graph.o:/usr/src/packages/BUILD/pd-l2ork/src/pd-l2ork-20190812+git1771+42cedfa8/pd/src/g_canvas.h:635: multiple definition of `array_joc'; g_canvas.o:/usr/src/packages/BUILD/pd-l2ork/src/pd-l2ork-20190812+git1771+42cedfa8/pd/src/g_canvas.h:635: first defined here
[  382s] /usr/bin/ld: g_graph.o:/usr/src/packages/BUILD/pd-l2ork/src/pd-l2ork-20190812+git1771+42cedfa8/pd/src/s_stuff.h:21: multiple definition of `pd_extrapath'; g_canvas.o:/usr/src/packages/BUILD/pd-l2ork/src/pd-l2ork-20190812+git1771+42cedfa8/pd/src/s_stuff.h:21: first defined here
... many more messages like these ...
[  383s] collect2: error: ld returned 1 exit status
[  383s] make[2]: *** [makefile:185: ../bin/pd-l2ork] Error 1
[  383s] make[2]: Leaving directory '/usr/src/packages/BUILD/pd-l2ork/src/pd-l2ork-20190812+git1771+42cedfa8/pd/src'
~~~

The full buildlog can be found at the OBS here: https://build.opensuse.org/build/home:aggraef/Arch/x86_64/pd-l2ork/_log. (I was building an older revision there, but it's the same error with the current HEAD.)

If you sift through the commit log of purr-data, you can probably find Jonathan's fixes and apply something similar here.

--- a/.gitmodules
+++ b/.gitmodules
@@ -16,3 +16,6 @@
 [submodule "Gem"]
 	path = Gem
 	url = https://git.purrdata.net/aggraef/gem.git
+[submodule "l2ork_addons/raspberry_pi/disis_gpio/wiringPi"]
+	path = l2ork_addons/raspberry_pi/disis_gpio/wiringPi
+	url = https://git.purrdata.net/jwilkes/wiringPi.git

--- a/externals/iem/iem_adaptfilt/src/iemlib.h
+++ b/externals/iem/iem_adaptfilt/src/iemlib.h
@@ -22,11 +22,11 @@ iemlib written by Thomas Musil, Copyright (c) IEM KUG Graz Austria 2000 - 2009 *
  */
 
 #ifdef MSW
-int sys_noloadbang;
+//int sys_noloadbang;
 //t_symbol *iemgui_key_sym=0;
 #include <io.h>
 #else
-extern int sys_noloadbang;
+//extern int sys_noloadbang;
 //extern t_symbol *iemgui_key_sym;
 #include <unistd.h>
 #endif

--- a/externals/iem/iem_ambi/src/iemlib.h
+++ b/externals/iem/iem_ambi/src/iemlib.h
@@ -22,11 +22,11 @@ iemlib written by Thomas Musil, Copyright (c) IEM KUG Graz Austria 2000 - 2009 *
  */
 
 #ifdef MSW
-int sys_noloadbang;
+//int sys_noloadbang;
 //t_symbol *iemgui_key_sym=0;
 #include <io.h>
 #else
-extern int sys_noloadbang;
+//extern int sys_noloadbang;
 //extern t_symbol *iemgui_key_sym;
 #include <unistd.h>
 #endif

--- a/externals/iem/iem_bin_ambi/src/iemlib.h
+++ b/externals/iem/iem_bin_ambi/src/iemlib.h
@@ -22,11 +22,11 @@ iemlib written by Thomas Musil, Copyright (c) IEM KUG Graz Austria 2000 - 2009 *
  */
 
 #ifdef MSW
-int sys_noloadbang;
+//int sys_noloadbang;
 //t_symbol *iemgui_key_sym=0;
 #include <io.h>
 #else
-extern int sys_noloadbang;
+//extern int sys_noloadbang;
 //extern t_symbol *iemgui_key_sym;
 #include <unistd.h>
 #endif

--- a/externals/iem/iem_delay/src/iemlib.h
+++ b/externals/iem/iem_delay/src/iemlib.h
@@ -22,11 +22,11 @@ iemlib written by Thomas Musil, Copyright (c) IEM KUG Graz Austria 2000 - 2009 *
  */
 
 #ifdef MSW
-int sys_noloadbang;
+//int sys_noloadbang;
 //t_symbol *iemgui_key_sym=0;
 #include <io.h>
 #else
-extern int sys_noloadbang;
+//extern int sys_noloadbang;
 //extern t_symbol *iemgui_key_sym;
 #include <unistd.h>
 #endif

--- a/externals/iem/iem_roomsim/src/iemlib.h
+++ b/externals/iem/iem_roomsim/src/iemlib.h
@@ -22,11 +22,11 @@ iemlib written by Thomas Musil, Copyright (c) IEM KUG Graz Austria 2000 - 2009 *
  */
 
 #ifdef MSW
-int sys_noloadbang;
+//int sys_noloadbang;
 //t_symbol *iemgui_key_sym=0;
 #include <io.h>
 #else
-extern int sys_noloadbang;
+//extern int sys_noloadbang;
 //extern t_symbol *iemgui_key_sym;
 #include <unistd.h>
 #endif

--- a/externals/iem/iem_spec2/src/iemlib.h
+++ b/externals/iem/iem_spec2/src/iemlib.h
@@ -22,11 +22,11 @@ iemlib written by Thomas Musil, Copyright (c) IEM KUG Graz Austria 2000 - 2009 *
  */
 
 #ifdef MSW
-int sys_noloadbang;
+//int sys_noloadbang;
 //t_symbol *iemgui_key_sym=0;
 #include <io.h>
 #else
-extern int sys_noloadbang;
+//extern int sys_noloadbang;
 //extern t_symbol *iemgui_key_sym;
 #include <unistd.h>
 #endif

--- a/externals/iem/iem_tab/src/iemlib.h
+++ b/externals/iem/iem_tab/src/iemlib.h
@@ -22,11 +22,11 @@ iemlib written by Thomas Musil, Copyright (c) IEM KUG Graz Austria 2000 - 2009 *
  */
 
 #ifdef MSW
-int sys_noloadbang;
+//int sys_noloadbang;
 //t_symbol *iemgui_key_sym=0;
 #include <io.h>
 #else
-extern int sys_noloadbang;
+//extern int sys_noloadbang;
 //extern t_symbol *iemgui_key_sym;
 #include <unistd.h>
 #endif

--- a/externals/iem/iemgui/src/iemlib.h
+++ b/externals/iem/iemgui/src/iemlib.h
@@ -22,11 +22,11 @@ iemlib written by Thomas Musil, Copyright (c) IEM KUG Graz Austria 2000 - 2009 *
  */
 
 #ifdef MSW
-int sys_noloadbang;
+//int sys_noloadbang;
 //t_symbol *iemgui_key_sym=0;
 #include <io.h>
 #else
-extern int sys_noloadbang;
+//extern int sys_noloadbang;
 //extern t_symbol *iemgui_key_sym;
 #include <unistd.h>
 #endif

--- a/l2ork_addons/raspberry_pi/disis_gpio/build.sh
+++ b/l2ork_addons/raspberry_pi/disis_gpio/build.sh
@@ -1,7 +1,5 @@
 curdir=`pwd`
-cd wiringPi/
-git reset --hard 96344ff7125182989f98d3be8d111952a8f74e15
-cd wiringPi/ && make static
+cd wiringPi/ && cd wiringPi/ && make static
 cd $curdir
 make
 exit 0

--- a/l2ork_addons/tar_em_up.sh
+++ b/l2ork_addons/tar_em_up.sh
@@ -363,8 +363,8 @@ requires internet connection to pull sources from various repositories..."
 	# copy rjlib into the extra folder
 	cp -rf rjlib ../packages/linux_make/build$inst_dir/lib/pd-l2ork/extra
 	# install raspberry pi externals (if applicable)
-	if [ $rpi -eq 1 ]
-	then
+	#if [ $rpi -eq 1 ]
+	#then
 		cd raspberry_pi
 		./makeall.sh
 		cp -f disis_gpio/disis_gpio.pd_linux ../../packages/linux_make/build$inst_dir/lib/pd-l2ork/extra
@@ -372,7 +372,7 @@ requires internet connection to pull sources from various repositories..."
 		cp -f disis_spi/disis_spi.pd_linux ../../packages/linux_make/build$inst_dir/lib/pd-l2ork/extra
 		cp -f disis_spi/disis_spi-help.pd ../../packages/linux_make/build$inst_dir/lib/pd-l2ork/extra
 		cd ../
-	fi
+	#fi
 	# install rtcmix~ external
 	#cd rtcmix-in-pd/
 	#git submodule update

--- a/l2ork_addons/tar_em_up.sh
+++ b/l2ork_addons/tar_em_up.sh
@@ -363,8 +363,8 @@ requires internet connection to pull sources from various repositories..."
 	# copy rjlib into the extra folder
 	cp -rf rjlib ../packages/linux_make/build$inst_dir/lib/pd-l2ork/extra
 	# install raspberry pi externals (if applicable)
-	#if [ $rpi -eq 1 ]
-	#then
+	if [ $rpi -eq 1 ]
+	then
 		cd raspberry_pi
 		./makeall.sh
 		cp -f disis_gpio/disis_gpio.pd_linux ../../packages/linux_make/build$inst_dir/lib/pd-l2ork/extra
@@ -372,7 +372,7 @@ requires internet connection to pull sources from various repositories..."
 		cp -f disis_spi/disis_spi.pd_linux ../../packages/linux_make/build$inst_dir/lib/pd-l2ork/extra
 		cp -f disis_spi/disis_spi-help.pd ../../packages/linux_make/build$inst_dir/lib/pd-l2ork/extra
 		cd ../
-	#fi
+	fi
 	# install rtcmix~ external
 	#cd rtcmix-in-pd/
 	#git submodule update

--- a/pd/src/g_canvas.c
+++ b/pd/src/g_canvas.c
@@ -47,6 +47,13 @@ t_canvas *canvas_editing;           /* last canvas to start text edting */
 t_canvas *canvas_whichfind;         /* last canvas we did a find in */ 
 //t_canvas *canvas_list;              /* list of all root canvases */
 
+// Ico, please review: I'm not sure whether these belong here, but they need
+// to be defined somewhere (in a .c file, not .h!!)
+t_garray* array_garray;
+t_class *preset_hub_class;
+t_class *preset_node_class;
+int array_joc;
+
 /* ------------------ forward function declarations --------------- */
 static void canvas_start_dsp(void);
 static void canvas_stop_dsp(void);

--- a/pd/src/g_canvas.h
+++ b/pd/src/g_canvas.h
@@ -151,7 +151,7 @@ typedef struct _arrayvis
     t_garray *av_garray;            /* owning structure */    
 } t_arrayvis;
 
-t_garray* array_garray;				/* used for sending bangs when
+EXTERN t_garray* array_garray;				/* used for sending bangs when
 									   array is changed  via gui */
 
 /* the t_tick structure describes where to draw x and y "ticks" for a glist */
@@ -235,8 +235,8 @@ struct _glist
 // this is where all the classes capable of being controlled via preset should be defined
 
 // preset objects
-t_class *preset_hub_class;
-t_class *preset_node_class;
+EXTERN t_class *preset_hub_class;
+EXTERN t_class *preset_node_class;
 
 // special case objects
 extern t_class *print_class;
@@ -632,7 +632,7 @@ EXTERN void array_resize(t_array *x, int n);
 EXTERN void array_free(t_array *x);
 EXTERN void array_redraw(t_array *a, t_glist *glist);
 EXTERN void array_resize_and_redraw(t_array *array, t_glist *glist, int n);
-int array_joc; /* for "jump on click" array inside a graph */
+EXTERN int array_joc; /* for "jump on click" array inside a graph */
 
 /* --------------------- gpointers and stubs ---------------- */
 EXTERN t_gstub *gstub_new(t_glist *gl, t_array *a);

--- a/pd/src/s_audio_mmio.c
+++ b/pd/src/s_audio_mmio.c
@@ -32,7 +32,6 @@ int nt_realdacblksize;
 #define MAXBUFFER 100   /* number of buffers in use at maximum advance */
 #define DEFBUFFER 30    /* default is about 30x6 = 180 msec! */
 static int nt_naudiobuffer = DEFBUFFER;
-float sys_dacsr = DEFAULTSRATE;
 
 static int nt_whichapi = API_MMIO;
 static int nt_meters;        /* true if we're metering */
@@ -127,6 +126,8 @@ int mmio_do_open_audio(void)
     if (sys_verbose)
         post("%d devices in, %d devices out",
             nt_nwavein, nt_nwaveout);
+
+    if (sys_dacsr == 0) sys_dacsr = DEFAULTSRATE;
 
     form.wf.wFormatTag = WAVE_FORMAT_PCM;
     form.wf.nChannels = CHANNELS_PER_DEVICE;
@@ -485,11 +486,6 @@ void nt_logerror(int which)
     }
 #endif
 }
-
-/* system buffer with t_sample types for one tick */
-t_sample *sys_soundout;
-t_sample *sys_soundin;
-float sys_dacsr;
 
 int mmio_send_dacs(void)
 {

--- a/pd/src/s_audio_oss.c
+++ b/pd/src/s_audio_oss.c
@@ -70,11 +70,6 @@ static t_oss_dev linux_adcs[OSS_MAXDEV];
 static int linux_noutdevs = 0;
 static int linux_nindevs = 0;
 
-    /* exported variables */
-t_float sys_dacsr;
-t_sample *sys_soundout;
-t_sample *sys_soundin;
-
     /* OSS-specific private variables */
 static int oss_blockmode = 1;   /* flag to use "blockmode"  */
 static char ossdsp[] = "/dev/dsp%d"; 

--- a/pd/src/s_path.c
+++ b/pd/src/s_path.c
@@ -43,6 +43,7 @@ t_namelist *sys_searchpath;
 t_namelist *sys_staticpath;
 t_namelist *sys_helppath;
 
+t_namelist *pd_extrapath;
 
     /* change '/' characters to the system's native file separator */
 void sys_bashfilename(const char *from, char *to)

--- a/pd/src/s_stuff.h
+++ b/pd/src/s_stuff.h
@@ -18,7 +18,7 @@ typedef struct _namelist    /* element in a linked list of stored strings */
     char *nl_string;            /* the string */
 } t_namelist;
 
-t_namelist *pd_extrapath;
+extern t_namelist *pd_extrapath;
 t_namelist *namelist_append(t_namelist *listwas, const char *s, int allowdup);
 t_namelist *namelist_append_files(t_namelist *listwas, const char *s);
 void namelist_free(t_namelist *listwas);

